runtime/connectors/blob/csvreader.go

```diff
@@ -0,0 +1,105 @@
+package blob
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"os"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"gocloud.dev/blob"
+)
+
+var _newLineSeparator = []byte("\n")
+
+func downloadCSV(ctx context.Context, bucket *blob.Bucket, obj *blob.ListObject, option *extractOption, fw *os.File) error {
+	reader := NewBlobObjectReader(ctx, bucket, obj)
+
+	rows, err := csvRows(reader, option)
+	if err != nil {
+		return err
+	}
+
+	// write rows
+	for _, r := range rows {
+		if _, err := fw.Write(r); err != nil {
+			return err
+		}
+		if _, err := fw.Write(_newLineSeparator); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func csvRows(reader *ObjectReader, option *extractOption) ([][]byte, error) {
+	if option.strategy == runtimev1.Source_ExtractPolicy_STRATEGY_TAIL {
+		return csvRowsTail(reader, option)
+	}
+	return csvRowsHead(reader, option)
+}
+
+func csvRowsTail(reader *ObjectReader, option *extractOption) ([][]byte, error) {
+	header, err := getHeader(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	remBytes := int64(option.limitInBytes - uint64(len(header)))
+	if _, err := reader.Seek(0-remBytes, io.SeekEnd); err != nil {
+		return nil, err
+	}
+
+	p := make([]byte, remBytes)
+	if _, err := reader.Read(p); err != nil {
+		return nil, err
+	}
+
+	rows := bytes.Split(p, _newLineSeparator)
+	// remove first row (possibly incomplete)
+	rows = rows[1:]
+	// append header at start
+	return append([][]byte{header}, rows...), nil
+}
+
+func csvRowsHead(reader *ObjectReader, option *extractOption) ([][]byte, error) {
+	if _, err := reader.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	p := make([]byte, option.limitInBytes)
+	if _, err := reader.Read(p); err != nil {
+		return nil, err
+	}
+
+	rows := bytes.Split(p, _newLineSeparator)
+	// remove last row (possibly incomplete)
+	return rows[:len(rows)-1], nil
+}
```

Comment:
1. Agreed. We can reduce memory usage with a buffer. Should we tackle it in a separate PR? Let me know in case we want to handle this here itself. 
2. Yes we can reduce the memory usage. It also makes it easy to handle some edge cases like reading till last new line separator. I will change it in this PR itself.

Reply:
We can tackle in another PR. If it's easy to do, let's do it – else if it's less easy, then let's just put it in backlog